File: src/core/datatypes.rs
use chrono::{DateTime, Utc};
use derive_more::{Deref, Display};
use notion_client::objects::block::{Block as NotionBlock, BlockType};
use notion_client::objects::parent::Parent;
use serde::{Deserialize, Serialize};

/// The identifier for a Notion `Page`. This exists to distinguish
/// between `Page` and `Block` identifiers at compile-time.
#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Hash, Display, Deref)]
pub struct PageID(String);

impl PageID {
    #[must_use]
    pub fn new(id: String) -> Self {
        PageID(id)
    }
}

/// The identifier for a Notion `Block`. This exists to distinguish
/// between `Block` and `Page` identifiers at compile-time.
#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq, Hash, Display, Deref)]
pub struct BlockID(String);

impl BlockID {
    #[must_use]
    pub fn new(id: String) -> Self {
        BlockID(id)
    }
}

/// A Block represents a single unit of notetaking, and its structure is heavily borrowed
/// from the Notion API's [Block object](https://developers.notion.com/reference/block).
///
/// A Block is always contained within a Page. Its content is defined by the `block_type` field,
/// but a plain form of the Block's text is stored in the `text` field. Finally, a Block
/// can have children, but if you want to fetch them then you must use the
/// notion::retrieve_all_block_children function with this Block's ID
#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub struct Block {
    pub id: BlockID,
    pub page_id: PageID,
    pub block_type: BlockType,
    pub text: String,
    pub creation_date: DateTime<Utc>,
    pub update_date: DateTime<Utc>,
    pub parent: Option<Parent>,
    pub has_children: bool,
}

impl std::hash::Hash for Block {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.0.hash(state);
    }
}

impl Block {
    #[must_use]
    pub fn from_notion_block(notion_block: NotionBlock, page_id: String) -> Self {
        Block {
            id: BlockID(notion_block.id.unwrap_or_default()),
            // TODO: consider removing this, since it is stored multiple times
            // throughout all the blocks, and we don't need it specifically on a block
            // it's a nice-to-have right now
            page_id: PageID(page_id),
            block_type: notion_block.block_type.clone(),
            // this is where the actual Block data is
            // TODO: instead of losing the data about Notion backlinks, as we're currently doing,
            // we should return a more general object that retains the info about Notion backlinks
            text: notion_block
                .block_type
                // TODO: notion-client mushes all of the text of certain BlockTypes (NumberedListItem, BulletListItem, Toggle, ToDo,
                // maybe some others) into a single Vec<Option<String>>, which is not great. When there's a need we should go back here
                // and do our own, more markdown-friendly way of extractin text for the different BlockTypes
                .plain_text()
                .into_iter()
                .map(Option::unwrap_or_default)
                .collect::<Vec<String>>()
                .join(" "), // TODO: a space " " separator is not always appropriate, but works for now. Find a better way to join the text
            creation_date: notion_block.created_time.unwrap_or_default(),
            update_date: notion_block.last_edited_time.unwrap_or_default(),
            parent: notion_block.parent,
            has_children: notion_block.has_children.unwrap_or_default(),
        }
    }

    #[must_use]
    pub fn to_markdown(&self) -> String {
        match &self.block_type {
            BlockType::Heading1 { heading_1: _ } => format!("# {}", self.text),
            BlockType::Heading2 { heading_2: _ } => format!("## {}", self.text),
            BlockType::Heading3 { heading_3: _ } => format!("### {}", self.text),
            BlockType::BulletedListItem {
                bulleted_list_item: _,
            } => format!("- {}", self.text),
            BlockType::NumberedListItem {
                numbered_list_item: _,
            } => format!("1. {}", self.text),
            BlockType::ToDo { to_do: _ } => format!("- [ ] {}", self.text),
            BlockType::Toggle { toggle: _ } => format!("> {}", self.text),
            _ => format!("{}", self.text),
        }
    }

    #[inline]
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.text.is_empty()
    }
}

/// A Page is a container for Blocks, and its structure is heavily borrowed
/// from the Notion API's [Page object](https://developers.notion.com/reference/page).
///
/// The Page struct exists because the data sources (such as Notion) that we ingest
/// from all have the concept of a collection of Blocks, and provide APIs for fetching
/// recently edited collections of Blocks. We call those Pages.
#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub struct Page {
    pub id: PageID,
    pub title: String,
    pub url: String,
    pub creation_date: DateTime<Utc>,
    pub update_date: DateTime<Utc>,
    pub child_blocks: Vec<Block>,
}

#[cfg(test)]
mod tests {
    use notion_client::objects::{
        block::{BulletedListItemValue, TextColor},
        rich_text::{RichText, Text},
    };

    use super::*;

    #[test]
    fn test_block_to_markdown() {
        let blocks = vec![
            Block {
                id: BlockID("1".to_string()),
                block_type: BlockType::Heading1 {
                    heading_1: Default::default(),
                },
                text: "Heading 1".to_string(),
                creation_date: Utc::now(),
                update_date: Utc::now(),
                parent: None,
                has_children: false,
                page_id: PageID("7b1b3b0c-14cb-45a6-a4b6-d2b48faecccb".to_string()),
            },
            Block {
                id: BlockID("2".to_string()),
                block_type: BlockType::Heading2 {
                    heading_2: Default::default(),
                },
                text: "Heading 2".to_string(),
                creation_date: Utc::now(),
                update_date: Utc::now(),
                parent: None,
                has_children: false,
                page_id: PageID("7b1b3b0c-14cb-45a6-a4b6-d2b48faecccb".to_string()),
            },
            Block {
                id: BlockID("3".to_string()),
                block_type: BlockType::BulletedListItem {
                    bulleted_list_item: BulletedListItemValue {
                        rich_text: vec![RichText::Text {
                            plain_text: Some("Bullet point".to_string()),
                            href: None,
                            annotations: None,
                            text: Text {
                                content: "Bullet point".to_string(),
                                link: None,
                            },
                        }],
                        color: TextColor::Default,
                        children: None,
                    },
                },
                text: "Bullet point".to_string(),
                creation_date: Utc::now(),
                update_date: Utc::now(),
                parent: None,
                has_children: false,
                page_id: PageID("7b1b3b0c-14cb-45a6-a4b6-d2b48faecccb".to_string()),
            },
            Block {
                id: BlockID("4".to_string()),
                block_type: BlockType::Paragraph {
                    paragraph: Default::default(),
                },
                text: "Normal text".to_string(),
                creation_date: Utc::now(),
                update_date: Utc::now(),
                parent: None,
                has_children: false,
                page_id: PageID("7b1b3b0c-14cb-45a6-a4b6-d2b48faecccb".to_string()),
            },
        ];

        let expected_markdown = "# Heading 1\n## Heading 2\n- Bullet point\nNormal text";
        let result_markdown = blocks
            .iter()
            .map(|block| block.to_markdown())
            .collect::<Vec<String>>()
            .join("\n");

        assert_eq!(result_markdown, expected_markdown);
    }
}

File: server/lib/datatypes.ts
import { DateTime } from 'luxon';
import type { BlockObjectResponse, RichTextItemResponse } from '@notionhq/client/build/src/api-endpoints';

// Equivalent TypeScript types for the Rust datatypes

/**
 * The identifier for a Notion `Page`. This exists to distinguish
 * between `Page` and `Block` identifiers at compile-time.
 */
export class PageID {
  constructor(private id: string) {}

  toString(): string {
    return this.id;
  }
}

/**
 * The identifier for a Notion `Block`. This exists to distinguish
 * between `Block` and `Page` identifiers at compile-time.
 */
export class BlockID {
  constructor(private id: string) {}

  toString(): string {
    return this.id;
  }
}

/**
 * Type representing the different types of blocks in Notion
 */
export type BlockType = 
  | { type: 'paragraph', paragraph: any }
  | { type: 'heading_1', heading_1: any }
  | { type: 'heading_2', heading_2: any }
  | { type: 'heading_3', heading_3: any }
  | { type: 'bulleted_list_item', bulleted_list_item: any }
  | { type: 'numbered_list_item', numbered_list_item: any }
  | { type: 'to_do', to_do: any }
  | { type: 'toggle', toggle: any }
  | { type: 'code', code: any }
  | { type: 'callout', callout: any }
  | { type: 'quote', quote: any }
  | { type: 'image', image: any }
  | { type: 'bookmark', bookmark: any }
  | { type: 'divider', divider: any }
  | { type: 'table', table: any }
  | { type: 'column_list', column_list: any }
  | { type: 'column', column: any }
  | { type: 'table_of_contents', table_of_contents: any }
  | { type: 'embed', embed: any }
  | { type: 'video', video: any }
  | { type: 'file', file: any }
  | { type: 'pdf', pdf: any }
  | { type: 'synced_block', synced_block: any }
  | { type: 'template', template: any }
  | { type: 'link_preview', link_preview: any }
  | { type: 'unsupported', unsupported: any };

/**
 * Parent type for blocks (simplified from Notion API)
 */
export interface Parent {
  type: string;
  page_id?: string;
  database_id?: string;
  block_id?: string;
  workspace?: boolean;
}

/**
 * A Block represents a single unit of notetaking, and its structure is heavily borrowed
 * from the Notion API's Block object.
 *
 * A Block is always contained within a Page. Its content is defined by the `block_type` field,
 * but a plain form of the Block's text is stored in the `text` field. Finally, a Block
 * can have children, but you need to fetch them separately.
 */
export class Block {
  id: BlockID;
  page_id: PageID;
  block_type: BlockType;
  text: string;
  creation_date: DateTime;
  update_date: DateTime;
  parent?: Parent;
  has_children: boolean;

  constructor(
    id: BlockID,
    page_id: PageID,
    block_type: BlockType,
    text: string,
    creation_date: DateTime,
    update_date: DateTime,
    parent: Parent | undefined,
    has_children: boolean
  ) {
    this.id = id;
    this.page_id = page_id;
    this.block_type = block_type;
    this.text = text;
    this.creation_date = creation_date;
    this.update_date = update_date;
    this.parent = parent;
    this.has_children = has_children;
  }

  /**
   * Converts a Notion API BlockObjectResponse to our application's Block type
   */
  static fromNotionBlockResponse(blockResponse: BlockObjectResponse, pageId: string): Block {
    // Extract the plain text from the rich_text arrays based on block type
    const extractPlainText = (block: BlockObjectResponse): string => {
      const blockType = block.type;
      
      // Handle different block types to extract text
      switch (blockType) {
        case 'paragraph':
        case 'heading_1':
        case 'heading_2':
        case 'heading_3':
        case 'bulleted_list_item':
        case 'numbered_list_item':
        case 'quote':
        case 'to_do':
        case 'toggle':
        case 'callout':
          // @ts-ignore - TypeScript doesn't understand the discriminated union here
          return block[blockType].rich_text
            .map((rt: RichTextItemResponse) => rt.plain_text || '')
            .join(' ');
            
        case 'code':
          // @ts-ignore
          return block.code.rich_text
            .map((rt: RichTextItemResponse) => rt.plain_text || '')
            .join(' ');
            
        case 'image':
        case 'file':
        case 'pdf':
        case 'video':
        case 'embed':
        case 'bookmark':
          // @ts-ignore
          return block[blockType].caption ? 
            // @ts-ignore
            block[blockType].caption
              .map((rt: RichTextItemResponse) => rt.plain_text || '')
              .join(' ') : '';
              
        case 'table_row':
          // @ts-ignore
          return block.table_row.cells
            .map((cell: RichTextItemResponse[]) => 
              cell.map((rt: RichTextItemResponse) => rt.plain_text || '').join(' ')
            )
            .join(' | ');
            
        case 'equation':
          // @ts-ignore
          return block.equation.expression || '';
          
        case 'link_preview':
          // @ts-ignore
          return block.link_preview.url || '';
          
        case 'divider':
        case 'breadcrumb':
        case 'table_of_contents':
        case 'column_list':
        case 'column':
        case 'link_to_page':
        case 'synced_block':
        case 'template':
        case 'table':
        case 'unsupported':
          return ''; // These block types don't have direct text content
          
        default:
          return '';
      }
    };

    // Create the block type object
    const blockTypeObj: BlockType = {
      type: blockResponse.type,
      [blockResponse.type]: (blockResponse as any)[blockResponse.type]
    } as BlockType;

    return new Block(
      new BlockID(blockResponse.id),
      new PageID(pageId),
      blockTypeObj,
      extractPlainText(blockResponse),
      DateTime.fromISO(blockResponse.created_time),
      DateTime.fromISO(blockResponse.last_edited_time),
      blockResponse.parent as Parent,
      blockResponse.has_children
    );
  }

  /**
   * Converts the block to markdown format
   */
  toMarkdown(): string {
    switch (this.block_type.type) {
      case 'heading_1':
        return `# ${this.text}`;
      case 'heading_2':
        return `## ${this.text}`;
      case 'heading_3':
        return `### ${this.text}`;
      case 'bulleted_list_item':
        return `- ${this.text}`;
      case 'numbered_list_item':
        return `1. ${this.text}`;
      case 'to_do':
        return `- [ ] ${this.text}`;
      case 'toggle':
        return `> ${this.text}`;
      default:
        return this.text;
    }
  }

  /**
   * Checks if the block is empty
   */
  isEmpty(): boolean {
    return this.text.length === 0;
  }
}

/**
 * A Page is a container for Blocks, and its structure is heavily borrowed
 * from the Notion API's Page object.
 *
 * The Page struct exists because the data sources (such as Notion) that we ingest
 * from all have the concept of a collection of Blocks, and provide APIs for fetching
 * recently edited collections of Blocks. We call those Pages.
 */
export class Page {
  id: PageID;
  title: string;
  url: string;
  creation_date: DateTime;
  update_date: DateTime;
  child_blocks: Block[];

  constructor(
    id: PageID,
    title: string,
    url: string,
    creation_date: DateTime,
    update_date: DateTime,
    child_blocks: Block[]
  ) {
    this.id = id;
    this.title = title;
    this.url = url;
    this.creation_date = creation_date;
    this.update_date = update_date;
    this.child_blocks = child_blocks;
  }
}

/**
 * Converts a Block to Markdown format
 */
export function blockToMarkdown(block: Block): string {
  return block.toMarkdown();
}